benchmark/agbenchmark/__main__.py

```diff
@@ -54,245 +69,335 @@ def get_unique_categories() -> set[str]:
 
         if maintain:
-            print("Running only regression tests")
-            pytest_args.append("--maintain")
+            logger.info("Running only regression tests")
         elif improve:
-            print("Running only non-regression tests")
-            pytest_args.append("--improve")
+            logger.info("Running only non-regression tests")
         elif explore:
-            print("Only attempt challenges that have never been beaten")
-            pytest_args.append("--explore")
+            logger.info("Only attempt challenges that have never been beaten")
 
     if mock:
-        pytest_args.append("--mock")
+        # TODO: unhack
         os.environ[
             "IS_MOCK"
         ] = "True"  # ugly hack to make the mock work when calling from API
 
-    if no_dep:
-        pytest_args.append("--no_dep")
-
-    if nc and cutoff:
-        print(
-            "Error: You can't use both --nc and --cutoff at the same time. Please choose one."
-        )
-        return 1
+    # Pass through flags
+    for flag, active in {
+        "--maintain": maintain,
+        "--improve": improve,
+        "--explore": explore,
+        "--no-dep": no_dep,
+        "--mock": mock,
+        "--nc": no_cutoff,
+        "--keep-answers": keep_answers,
+    }.items():
+        if active:
+            pytest_args.append(flag)
 
-    if nc:
-        pytest_args.append("--nc")
     if cutoff:
-        pytest_args.append("--cutoff")
-        print(f"Setting cuttoff override to {cutoff} seconds.")
+        pytest_args.append(f"--cutoff={cutoff}")
+        logger.debug(f"Setting cuttoff override to {cutoff} seconds.")
+
     current_dir = Path(__file__).resolve().parent
-    print(f"Current directory: {current_dir}")
-    pytest_args.extend((str(current_dir), "--cache-clear"))
+    pytest_args.append(str(current_dir / "generate_test.py"))
+
+    pytest_args.append("--cache-clear")
     exit_code = pytest.main(pytest_args)
-    SingletonReportManager().clear_instance()
+
+    SingletonReportManager.clear_instance()
+    return exit_code
+
+
+def validate_args(
+    maintain: bool,
+    improve: bool,
+    explore: bool,
+    tests: Sequence[str],
+    categories: Sequence[str],
+    skip_categories: Sequence[str],
+    no_cutoff: bool,
+    cutoff: Optional[int],
+) -> None:
+    if (maintain + improve + explore) > 1:
+        raise InvalidInvocationError(
+            "You can't use --maintain, --improve or --explore at the same time. "
+            "Please choose one."
+        )
+
+    if tests and (categories or skip_categories or maintain or improve or explore):
+        raise InvalidInvocationError(
+            "If you're running a specific test make sure no other options are "
+            "selected. Please just pass the --test."
+        )
+
+    if no_cutoff and cutoff:
+        raise InvalidInvocationError(
+            "You can't use both --nc and --cutoff at the same time. "
+            "Please choose one."
+        )
```

Comment:
Does click have a built-in mutual exclusion flag for params or is this what that is?

Reply:
Not built-in. There are packages available, but this is easier.